File: go/Trapping_Rain_Water.go
package main

func trap(h []int) int {
	var stk []int
	mx := 0
	ans := 0
	for i := 0; i < len(h); i++ {
		if len(stk) == 0 {
			stk = append(stk, i)
			mx = i
			continue
		}

		if h[i] < h[mx] {
			stk = append(stk, i)
			continue
		}

		sub := 0
		for _, v := range stk {
			sub += h[v]
		}
		ans += (i-mx)*h[mx] - sub

		stk = []int{i}
		mx = i
	}

	if len(stk) > 1 {
		var try []int
		for i := len(stk) - 1; i >= 0; i-- {
			try = append(try, h[stk[i]])
		}
		ans += trap(try)
	}
	return ans
}

File: cpp/Trapping_Rain_Water.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
  int trap(vector<int> &h) {
    if (h.size() <= 0) {
      return 0;
    }
    int l = 0, r = h.size() - 1;
    int ml = h[l], mr = h[r], ans = 0;
    while (l < r) {
      if (ml < mr) {
        l++;
        ml = max(ml, h[l]);
        ans += ml - h[l];
      } else {
        r--;
        mr = max(mr, h[r]);
        ans += mr - h[r];
      }
    }
    return ans;
  }
};
